pytorch_lightning/core/lightning.py

```diff
@@ -1176,22 +1176,25 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
             optimizer: Current optimizer used in training_loop
             optimizer_idx: Current optimizer idx in training_loop
         """
+
+        # Iterate over all optimizer parameters to preserve their `requires_grad` information
+        # in case these are pre-defined during `configure_optimizers`
         param_requires_grad_state = {}
-        # make sure current optimizer is latest to be iterated over.
-        optimizers = [opt for opt in self.optimizers(use_pl_optimizer=False) if opt != optimizer] + [optimizer]
-        num_optimizers = len(optimizers) - 1
-        for opt_idx, opt in enumerate(optimizers):
+        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
             for group in opt.param_groups:
                 for param in group['params']:
-                    if num_optimizers == opt_idx:
-                        # If a param appears in 2 optimizers, revert `requires_grad` to before toggle.
-                        if param in param_requires_grad_state:
-                            param.requires_grad = param_requires_grad_state[param]
+                    # If a param appears in multiple optimizers, use the `requires_grad` info already set
+                    if param in param_requires_grad_state:
```

Comment:
Could we have a test with 3 optimizers and shared parameters ? I think the previous code made a 2 optimizers assumption, but this new code works for more.

The assumption was done there, where 3 optimizers could share the same parameters, and by setting  `param.requires_grad = False` with first optimizer, the second one will override the state and we lose the information.
```
param_requires_grad_state[param] = param.requires_grad		                         
param.requires_grad = False
```